Added cleaning pads interval

NEVdataDyne opened the pull request.

edenhaus commented on deebot_client/hardware/deebot/p1jij8.py:
```diff
@@ -120,6 +121,17 @@
+                get=[GetPadsCleaningInterval()],
+                set=SetPadsCleaningInterval,
+                types=(
+                    SetPadsCleaningInterval.DEEP,
```
`SetPadsCleaningInterval` doesn't exist. Did you commit and push all the code?

NEVdataDyne replied:
I wrote SetPadsCleaningInterval in deebot_client/commands/json/pads_cleaning_interval.py, do I need to add it somewhere else?

edenhaus commented on deebot_client/commands/json/pads_cleaning_interval.py:
```diff
@@ -0,0 +1,39 @@
+    get_command = GetPadsCleaningInterval
+    _mqtt_params = {"interval": InitParam(PadsCleaningInterval)}
+
+    def __init__(self, interval: PadsCleaningInterval | str) -> None:
+        if isinstance(interval, str):
+            interval = PadsCleaningInterval.get(interval)
+        super().__init__({"interval": interval.value})
```
Can you set the interval to any minute? E.g. to 7 minutes?

If yes, I would suggest you to refactor the the command like the volume commands, where the user can specify any number

NEVdataDyne replied:
Yes you can set any amount of minutes. I did want to develop the cleaning pads interval command to match the volume command however I couldn't because I used the code you merged on PR #449 for work mode and copied what you did without understanding it. I don't have such "delta" with the volume command so I am lost on so many levels...

edenhaus commented on deebot_client/events/__init__.py:
```diff
@@ -49,6 +50,8 @@
     "WaterInfoEvent",
     "WorkMode",
     "WorkModeEvent",
+    "PadsCleaningInterval",
```
Please sort it alphabetically.

edenhaus commented on deebot_client/commands/json/__init__.py:
```diff
@@ -150,7 +153,10 @@
-    SetWorkMode
+    SetWorkMode,
+
+    GetPadsCleaningInterval,
```
Please sort it alphabetically.

edenhaus commented on deebot_client/commands/json/__init__.py:
```diff
@@ -79,6 +80,8 @@
     "SetWaterInfo",
     "GetWorkMode",
     "SetWorkMode",
+    "GetPadsCleaningInterval",
```
Please sort it alphabetically.

edenhaus reviewed (changes requested):
thanks for the review and each PR helps :)
I have left some suggestions, could you please address them?

edenhaus commented:
I quickly fixed my requested changes. Can you please test if the commands are working?

NEVdataDyne commented:
Yes of course, I'll do it tomorrow. Thank you

NEVdataDyne commented:
Could you please tell me how to test? On the work mode pull request there was a zip file to install manually but I don't find it for this pull request, should I manually install the modified files instead?

edenhaus commented:
You did not open yet a PR to add these new commands to the HA integration. I currently have time to do it, so please open it your self.

Otherwise you can run the python code directly with the example in the readme

NEVdataDyne commented:
Ok, I have to do something like that also?
https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/pull/449/files

I'll try to do that tomorrow evening

edenhaus commented:
> Ok, I have to do something like that also? [DeebotUniverse/Deebot-4-Home-Assistant#449 (files)](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/pull/449/files)
> 
> I'll try to do that tomorrow evening

Yes but this time you need to add a new number entity not a select one

NEVdataDyne commented:
I did... something... I hope it is helpful
#457

NEVdataDyne commented:
Could you please help me create the zip file soI can test?

MVladislav commented:
Hi @NEVdataDyne

what i have done to test it fast in Home Assistant, is to switch the "**Deebot-4-Home-Assistant**" "Default branch", under github repo settings, to the current working branch, what would be for you from [`dev`](https://github.com/NEVdataDyne/Deebot-4-Home-Assistant) to [`Cln-pads-interval`](https://github.com/NEVdataDyne/Deebot-4-Home-Assistant/tree/Cln-pads-interval) and update the "**[manifest.json](https://github.com/NEVdataDyne/Deebot-4-Home-Assistant/blob/Cln-pads-interval/custom_components/deebot/manifest.json)**" by change the "requirements" from `"deebot-client==4.0.0b0",` for you than to `"git+https://github.com/NEVdataDyne/client.py.git@dev#deebot-client==4.0.0dev0",` (that HACS will load the correct client.py version you changed).

Afterward you can import this repo over **"HACS"** directly into Home Assistant by using `Custom repositories` where you can add your github url `https://github.com/NEVdataDyne/Deebot-4-Home-Assistant.git` and install your own version of "**Deebot-4-Home-Assistant**" for testings into Home Assistant (and if you make any changes in the git repo you can easy click on redownload over HACS).

(_when you later make a pull request, do not forget to revert the changes made in manifest.json_)

Hope this helps you somehow.

NEVdataDyne commented:
@MVladislav, thank you for your reply. I am trying to understand as much as possible what I am doing and how GitHub works.
1) I merged Cln-pads-interval with dev in "Deebot-4-Home-Assistant", as a result I didn't change the default branch to [Cln-pads-interval](https://github.com/NEVdataDyne/Deebot-4-Home-Assistant/tree/Cln-pads-interval)
2) In manifest.json, I changed the "requirements" to "git+https://github.com/NEVdataDyne/client.py.git@dev#deebot-client==4.0.0dev0". If I understand correctly, edenhaus did the corrections directly in my repository. How do I create this URL? Especially the "4.0.0dev0" part?
3) I added the integration using HACS, the problem with this method is that now I have two "Deebot 4 Home assistant" making things a bit unclear but I didn't have to use ssh so that's good.

When I go to Settings->Devices & Services->Integrations->Add integration and want to add Deebot-4-Home-Assistant, I get the error "Config flow could not be loaded: unexpected character: line 74 column 7 (char 2059)". I don't see anything wrong in custom_components/deebot/config_flow.py but I am not sure it refers to this file.

MVladislav commented:
Hi @NEVdataDyne 

> In manifest.json, I changed the "requirements" to "git+https://github.com/NEVdataDyne/client.py.git@dev#deebot-client==4.0.0dev0".
> If I understand correctly, edenhaus did the corrections directly in my repository. How do I create this URL? Especially the "4.0.0dev0" part?

To be honest, I not 100% understand why it is needed to add with line after `#`
But as I understand, before the `#` is the fallback version and as long the version after `#` is not official deployed it will load the git source fallback.
Without `#` it has not worked for me.

> I added the integration using HACS, the problem with this method is that now I have two "Deebot 4 Home assistant" making things a bit unclear but I didn't have to use ssh so that's good.

This should not make a problem, as the latest one you installed should overwrite the content you have installed before.
You can remove one of the version, to not always search which is which.
On remove, you will get a popup "Integration is configured" where you can click "IGNORE" and afterward you need to reinstall the other version, as the remove will remove the installed folder.

> When I go to Settings->Devices & Services->Integrations->Add integration and want to add Deebot-4-Home-Assistant,
> I get the error "Config flow could not be loaded: unexpected character: line 74 column 7 (char 2059)".
> I don't see anything wrong in custom_components/deebot/config_flow.py but I am not sure it refers to this file.

In [number.py#L86](https://github.com/NEVdataDyne/Deebot-4-Home-Assistant/blob/d3259cdb80f722b9cc75e0ef0f6ac69bd3d0ee5e/custom_components/deebot/number.py#L86) you defined as key `clean_pads_interval`,
but in the translation [en.json#L75](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/blob/d3259cdb80f722b9cc75e0ef0f6ac69bd3d0ee5e/custom_components/deebot/translations/en.json#L75) you add it as `interval`. Also a comma is missing before interval definition in json translation files.

I have add some comments in the [pull request](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/pull/457#pullrequestreview-1726234197).

NEVdataDyne commented:
Thank you for your help. I made the changes in my GitHub repository (didn't make a new pull request), then following your method I imported the repo in HACS and downloadedand the integration in my HA. Then I forced HA to reload it using schmoogol's command : https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/issues/434#issuecomment-1780628736

-> Now I understand that at no point in time I loaded two deebot integration but I simply overwrite the existing one.

A cleaning pads interval entity did appear in the list (disabled). After enabling it I changed its value but it has no effect on the robot's configuration. I also have this "not a number" label hinting at some issue...

![image](https://github.com/DeebotUniverse/client.py/assets/82730421/866758bc-b3e5-451f-860d-43bcb3830958)

NEVdataDyne commented:
About interval vs clean_pads_interval, I used interval because that's the name of the property in the json command file sent to Ecovac's servers. I don't know when the code in client.py refers to the json property or a variable so I thought I'd put interval everywhere.

Now I am wondering if somewhere in client.py I shoudl also rename interval to clean_pads_interval at some places...?

MVladislav commented on deebot_client/commands/json/pads_cleaning_interval.py:
```diff
@@ -0,0 +1,39 @@
+class SetPadsCleaningInterval(SetCommand):
+    """Set pads cleaning interval command."""
+
+    name = "setPadsCleaningInterval"
```
Is this command correct?
Is it related to your comment in [issue-1918161365](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/issues/434#issue-1918161365) in file [Ecovacs T20 Cleaning pads interval.txt](https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/files/12753366/Ecovacs.T20.Cleaning.pads.interval.txt)?
In the file the command is `setWashInterval` but here used `setPadsCleaningInterval`

NEVdataDyne replied:
It works!

I changed both the name in the set method and in the get method. Because I based my work on the work mode which had the name of the set function identical to the json cmdName, I didn't realize what this "name variable" was.

I still have a question though, for the get function, I put name = "getWashInterval" because it seemed to be the thing to do but I never saw a getWashInterval in a json, is it used? How one would know for sure what to put instead of simply guessing?

MVladislav replied:
nice :)

**get** and **set** are in all cases, I always saw, same named.
The **get** is always called when the app/screen with info is opened. 
When you not saw the **get** request directly, it could be it was not called in a single request, instead it was called in a multi **get** request, which you see in the **`getInfo`** request like this example:

```json
{
  "request": {
    "td": "q",
    "toId": "REPLACED",
    "toType": "p95mgv",
    "toRes": "REPLACED",
    "cmdName": "getInfo",
    "payloadType": "j",
    "auth": {
      "with": "REPLACED",
      "userid": "REPLACED",
      "realm": "REPLACED",
      "token": "REPLACED",
      "resource": "REPLACED"
    },
    "app": {
      "id": "ecovacs"
    },
    "payload": {
      "body": {
        "data": [
          "getWaterInfo",
          "getCleanPreference",
          "getStationInfo",
          "getAutoEmpty",
          "getSweepMode"
        ]
      },
      "header": {
        "pri": 2,
        "ts": "1699216144028",
        "tzm": 60,
        "ver": "0.0.22"
      }
    }
  },
  "response": {
    "id": "REPLACED",
    "ret": "ok",
    "resp": {
      "header": {
        "pri": 1,
        "tzm": 60,
        "ts": "1699216136655",
        "ver": "0.0.1",
        "fwVer": "1.7.2",
        "hwVer": "0.1.1",
        "wkVer": "0.1.54"
      },
      "body": {
        "code": 0,
        "msg": "ok",
        "data": {
          "getWaterInfo": {
            "code": 0,
            "msg": "ok",
            "data": {
              "enable": 0,
              "amount": 2,
              "type": 0,
              "sweepType": 1
            }
          },
          "getCleanPreference": {
            "data": {
              "enable": 1
            },
            "code": 0,
            "msg": "ok"
          },
          "getStationInfo": {
            "data": {
              "state": 1,
              "name": "Empty Station",
              "model": "AES",
              "sn": "317211735002292000387",
              "wkVer": "0.1.3"
            },
            "code": 0,
            "msg": "OK"
          },
          "getAutoEmpty": {
            "code": 0,
            "msg": "OK",
            "data": {
              "status": 0,
              "enable": 1,
              "frequency": "auto"
            }
          }
        }
      }
    },
    "payloadType": "j"
  }
}
```

NEVdataDyne replied:
Ah yes thanks, I did see such json but didn't pay much attention because I didn't think they would be usefull. I'll add the one for T20 in change request #434

edenhaus commented:
> About interval vs clean_pads_interval, I used interval because that's the name of the property in the json command file sent to Ecovac's servers. I don't know when the code in client.py refers to the json property or a variable so I thought I'd put interval everywhere.
> 
> Now I am wondering if somewhere in client.py I shoudl also rename interval to clean_pads_interval at some places...?

I would prefer `clean_pads_interval` over `interval` as interval could be anything.

edenhaus commented:
Is this PR tested?

NEVdataDyne commented:
Yes it is working on my HA

NEVdataDyne commented on deebot_client/commands/json/pads_cleaning_interval.py:
```diff
@@ -11,29 +12,28 @@
         return HandlingResult.success()
 
 
 class SetPadsCleaningInterval(SetCommand):
```
I chose PadsCleaningInterval for readibility, if it changes to Wash it will match the name Ecovacs uses but then it will be unclear wat it does don't you think?

Why is PadsCleaningInterval() not needed in PadsCleaningIntervalEvent()?

edenhaus replied:
The commands should have the same name as the ecovacs commands without any typos (they have made some). In the HA integration we can rename them if required but the commands in the library should follow ecovacs

Please change also the name of the event to match the new name

NEVdataDyne commented on deebot_client/commands/json/pads_cleaning_interval.py:
```diff
@@ -11,29 +12,28 @@
 class GetPadsCleaningInterval(CommandWithMessageHandling, MessageBodyDataDict):
```
What do you mean by docs? The file names?

edenhaus replied:
Don't mark the conversation as resolved, if you asked a question.

Also the file name and the python doc string need to be adjusted

NEVdataDyne commented:
@edenhaus Do you want me to change all strings from PadsCleaning and pads_cleaning to wash?

edenhaus commented on deebot_client/commands/json/pads_cleaning_interval.py:
```diff
@@ -0,0 +1,39 @@
+        return HandlingResult.success()
+
+
+class SeWashInterval(SetCommand):
```
```suggestion
class SetWashInterval(SetCommand):
```
